File: app/src/main/java/com/shahin/movieapp/network/ItemFetcher.java
package com.shahin.movieapp.network;

import android.util.Log;

import androidx.lifecycle.MutableLiveData;

import com.shahin.movieapp.model.Movie;
import com.shahin.movieapp.model.MoviesList;

import java.util.HashMap;
import java.util.Map;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

public class ItemFetcher {

    private static final String BASE_URL = "https://www.omdbapi.com/";
    private static ItemFetcher mInstance;
    private Map<String, String> mQueries;
    private static Retrofit mRetrofit;
    private ApiService mApiService;
    private MutableLiveData<MoviesList> mMutableLDMoviesList = new MutableLiveData<>();
    private MutableLiveData<Movie> mMutableLDMovie = new MutableLiveData<>();
    private static final String TAG = "Movie";

    private ItemFetcher() {
        mRetrofit = new Retrofit.Builder()
                .baseUrl(BASE_URL)
                .addConverterFactory(GsonConverterFactory.create())
                .build();
        mApiService = mRetrofit.create(ApiService.class);

        mQueries = new HashMap<String, String>() {{
            put("apikey", "3e974fca");
        }};
    }

    public static ItemFetcher getInstance() {
        if (mInstance == null) {
            mInstance = new ItemFetcher();
        }
        return mInstance;
    }

    public void getMoviesList() {
        Map<String, String> queries = new HashMap<>();
        queries.putAll(mQueries);
        queries.put("s", "batman");

        mApiService.getMoviesList(queries).enqueue(new Callback<MoviesList>() {
            @Override
            public void onResponse(Call<MoviesList> call, Response<MoviesList> response) {
                if (response.isSuccessful()) {
                    Log.d(TAG, "Successful getList ");
                    MoviesList moviesList = response.body();
                    mMutableLDMoviesList.postValue(moviesList);
                }
            }

            @Override
            public void onFailure(Call<MoviesList> call, Throwable t) {
                Log.d(TAG, "Failed getList : " + t.getMessage());
            }
        });

    }

    public void getMovie(String imdbID){
        Map<String,String> queries = new HashMap<>();
        queries.putAll(mQueries);
        queries.put("i",imdbID);

        mApiService.getMovie(queries).enqueue(new Callback<Movie>() {
            @Override
            public void onResponse(Call<Movie> call, Response<Movie> response) {
                if(response.isSuccessful()){
                    Log.d(TAG, "Successful getMovie ");
                    Movie movie = response.body();
                    mMutableLDMovie.postValue(movie);
                }
            }

            @Override
            public void onFailure(Call<Movie> call, Throwable t) {
                Log.d(TAG, "Failed getList " + t.getMessage());
            }
        });

    }

    public MutableLiveData<MoviesList> getMutableLDMoviesList() {
        return mMutableLDMoviesList;
    }

    public void setMutableLDMoviesList(MutableLiveData<MoviesList> mutableLDMoviesList) {
        mMutableLDMoviesList = mutableLDMoviesList;
    }

    public MutableLiveData<Movie> getMutableLDMovie() {
        return mMutableLDMovie;
    }

    public void setMutableLDMovie(MutableLiveData<Movie> mutableLDMovie) {
        mMutableLDMovie = mutableLDMovie;
    }
}

File: app/src/main/java/com/shahin/movieapp/data/network/ItemFetcher.kt
package com.shahin.movieapp.data.network

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import com.shahin.movieapp.data.local.model.MovieDto
import com.shahin.movieapp.data.network.model.MoviesListDto
import com.shahin.movieapp.data.network.utils.BASE_URL
import com.shahin.movieapp.data.repository.toDomain
import com.shahin.movieapp.model.Movie
import com.shahin.movieapp.model.MovieItemList
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory


class ItemFetcher {

    private var queryParam: Map<String, String>
    private var retrofit: Retrofit
    private var movieService: MovieService

    companion object {
        private val itemFetcher: ItemFetcher by lazy { ItemFetcher() }
        fun getInstance(): ItemFetcher {
            return itemFetcher
        }
    }

    init {
        queryParam = mapOf("apikey" to "3e974fca")
        retrofit = Retrofit.Builder()
            .baseUrl(BASE_URL)
            .addConverterFactory(GsonConverterFactory.create())
            .build()

        movieService = retrofit.create(MovieService::class.java)

    }

    fun getMoviesList(): LiveData<List<MovieItemList>> {
        val queries = mutableMapOf<String, String>()
        queries.putAll(queryParam)
        queries["s"] = "batman"

        val moviesList = MutableLiveData<List<MovieItemList>>()
        movieService.getMoviesList(queries).enqueue(object : Callback<MoviesListDto> {
            override fun onResponse(call: Call<MoviesListDto>, response: Response<MoviesListDto>) {
                if (response.isSuccessful) {
                    moviesList.postValue(response.body()?.toDomain() ?: emptyList())
                }
            }

            override fun onFailure(call: Call<MoviesListDto>, t: Throwable) {

            }


        })

        return moviesList
    }

    fun getMovie(imdbID: String): LiveData<Movie?> {
        val queries = mutableMapOf<String, String>()
        queries.putAll(queryParam)
        queries["i"] = imdbID

        val movie = MutableLiveData<Movie?>()
        movieService.getMovie(queries).enqueue(object : Callback<MovieDto> {

            override fun onResponse(call: Call<MovieDto>, response: Response<MovieDto>) {
                if (response.isSuccessful) {
                    movie.postValue(response.body()?.toDomain())
                }
            }

            override fun onFailure(call: Call<MovieDto>, t: Throwable) {

            }
        })

        return movie

    }

}